bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java

```diff
@@ -1040,18 +1045,83 @@ private static boolean checkIsArrayType(BType sourceType, BArrayType targetType,
     }
 
     private static boolean checkIsTupleType(BType sourceType, BTupleType targetType, List<TypePair> unresolvedTypes) {
-        if (sourceType.getTag() != TypeTags.TUPLE_TAG) {
+
+        if (sourceType.getTag() == TypeTags.UNION_TAG) {
+            for (BType memberType : ((BUnionType) sourceType).getMemberTypes()) {
+                if (!checkIsTupleType(memberType, targetType, unresolvedTypes)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        if (sourceType.getTag() != TypeTags.ARRAY_TAG && sourceType.getTag() != TypeTags.TUPLE_TAG) {
+            return false;
+        }
+
+        List<BType> targetTypes = new ArrayList<>(targetType.getTupleTypes());
+        BType targetRestType = targetType.getRestType();
+
+        BTupleType sourceTupleType;
+        if (sourceType.getTag() == TypeTags.TUPLE_TAG) {
+            sourceTupleType = (BTupleType) sourceType;
+        } else {
+            BArrayType sourceArrayType = (BArrayType) sourceType;
+            BType sourceElementType = sourceArrayType.getElementType();
+
+            switch (sourceArrayType.getState()) {
+                case UNSEALED:
+                    if (targetRestType == null) {
+                        return false;
+                    } else {
+                        return checkIsType (sourceElementType, targetRestType, unresolvedTypes);
+                    }
```

Comment:
```suggestion
                    }
                    return checkIsType (sourceElementType, targetRestType, unresolvedTypes);
```

Reply:
Need to check non-rest member types of the target tuple too.

For example where `x` is `boolean[] x = [false, true];`, `<any> x is [string, boolean...]` evaluates to true now, but it shouldn't.

Basically an open array cannot belong to a tuple type unless it has a rest descriptor AND has no non-rest member types.